source/fe/mapping_fe.cc

```diff
@@ -0,0 +1,1385 @@
+
+  switch (mapping_type)
+    {
+    case mapping_covariant:
+    {
+      Assert (data.update_flags & update_contravariant_transformation,
+              typename FEValuesBase<dim>::ExcAccessToUninitializedField("update_contravariant_transformation"));
+
+      for (unsigned int i=0; i<output.size(); ++i)
+        output[i] = apply_transformation(data.covariant[i], input[i]);
+
+      return;
+    }
+    default:
+      Assert(false, ExcNotImplemented());
+    }
+
+}
+
+
+
+template<int dim, int spacedim, class DH, class VECTOR>
+Point<spacedim>
+MappingFE<dim,spacedim,DH,VECTOR>::
+transform_unit_to_real_cell (const typename Triangulation<dim,spacedim>::cell_iterator &cell,
+                             const Point<dim>                                 &p) const
+{
+//  Use the get_data function to create an InternalData with data vectors of
+//  the right size and transformation shape values already computed at point
+//  p. Make sure only one processor at a time performs this operations.
+
+  Threads::Mutex::ScopedLock lock(mutex);
+  update_internal_dofs(cell);
+  const Quadrature<dim> point_quadrature(p);
+  std::auto_ptr<InternalData>
+  mdata (dynamic_cast<InternalData *> (
+           get_data(update_transformation_values, point_quadrature)));
+
+  return this->transform_unit_to_real_cell_internal(*mdata);
+}
+
+
+template<int dim, int spacedim, class DH, class VECTOR>
+Point<spacedim>
+MappingFE<dim,spacedim,DH,VECTOR>::
+transform_unit_to_real_cell_internal (const InternalData &data) const
+{
+  Point<spacedim> p_real;
+
+  for (unsigned int i=0; i<data.n_shape_functions; ++i)
+    {
+      unsigned int comp_i = fe->system_to_component_index(i).first;
+      if (fe_mask[comp_i])
+        p_real[fe_to_real[comp_i]] += local_dofs[i] * data.shape(0,i);
+    }
+
+  return p_real;
+}
+
+
+
+template<int dim, int spacedim, class DH, class VECTOR>
+Point<dim>
+MappingFE<dim,spacedim,DH,VECTOR>::
+transform_real_to_unit_cell (const typename Triangulation<dim,spacedim>::cell_iterator &cell,
+                             const Point<spacedim>                            &p) const
+{
+  {
+    Threads::Mutex::ScopedLock lock(mutex);
+    update_internal_dofs(cell);
+  }
+  // first a Newton iteration based on the real mapping. It uses the center
+  // point of the cell as a starting point
+  Point<dim> initial_p_unit;
+
+
+  for (unsigned int d=0; d<dim; ++d)
+    initial_p_unit[d] = 0.5;
+
+  // use the full mapping. in case the function above should have given us
+  // something back that lies outside the unit cell (that might happen
+  // because we may have given a point 'p' that lies inside the cell with
+  // the higher order mapping, but outside the Q1-mapped reference cell),
+  // then project it back into the reference cell in hopes that this gives
+  // a better starting point to the following iteration
+  initial_p_unit = GeometryInfo<dim>::project_to_unit_cell(initial_p_unit);
```

Comment:
I don't understand the code here. Above you set `initial_p_unit` to (0.5,0.5). Why do you immediately need to project it back? It's already inside the reference cell.